src/MaxLikelihoodFit.cc

```diff
@@ -481,9 +483,51 @@ void MaxLikelihoodFit::getShapesAndNorms(RooAbsPdf *pdf, const RooArgSet &obs, s
             ns.signal  = (coeff->getStringAttribute("combine.process") ? coeff->getAttribute("combine.signal") : (strstr(ns.norm->GetName(),"shapeSig") != 0));
             std::auto_ptr<RooArgSet> myobs(ns.pdf->getObservables(obs));
             ns.obs.add(*myobs);
+            ns.isfunc = false;
         }
         return;
     }
+    RooRealSumPdf *sum = dynamic_cast<RooRealSumPdf *>(pdf);
+    if (sum != 0) {
+      RooArgList clist(sum->coefList());
+      RooArgList plist(sum->funcList());
+      if (plist.getSize() == 1) {
+         CMSHistErrorPropagator *err = dynamic_cast<CMSHistErrorPropagator*>(plist.at(0));
+         if (err) {
+           clist.removeAll();
+           plist.removeAll();
+           clist.add(err->coefList());
+           plist.add(err->wrapperList());
+         }
+      }
+      for (int i = 0, n = clist.getSize(); i < n; ++i) {
+        RooAbsReal *coeff = (RooAbsReal *) clist.at(i);
+        std::string coeffName = coeff->GetName();
+        if (coeffName.find(filterString_) == std::string::npos) continue;
+        ShapeAndNorm &ns = out[coeffName];
+        RooAbsReal* shape = (RooAbsReal*)plist.at(i);
+        std::auto_ptr<RooArgSet> myobs(shape->getObservables(obs));
+        ns.pdf = shape;
+        TString normProdName = TString::Format("%s_mlf_fullnorm", coeff->GetName());
+        RooAbsReal * normProd = nullptr;
+        if (coeff->ownedComponents()) {
+          normProd = dynamic_cast<RooAbsReal*>(coeff->ownedComponents()->find(normProdName));
+        }
+        if (!normProd) {
+          RooAbsReal* integral = shape->createIntegral(*myobs);
+          normProd = new RooProduct(normProdName, "", RooArgList(*integral, *coeff));
+          normProd->addOwnedComponents(RooArgSet(*integral));
+          coeff->addOwnedComponents(RooArgSet(*normProd));
+        }
+
+        ns.norm = normProd;
```

Comment:
Some gymnastics here to save the norm. Unlike for a pdf the norm is integral(function) * coefficient.